File: 486. Predict the Winner.py
class Solution:
    def PredictTheWinner(self, nums):
        n = len(nums)

        @lru_cache(None)
        def dfs(i,j):
            if i > j: return 0
            return max(nums[i]-dfs(i+1,j),nums[j]-dfs(i,j-1))

        return dfs(0,n-1) >= 0

File: 486. Predict the Winner.rs
use std::cmp;

impl Solution {
    pub fn predict_the_winner(nums: Vec<i32>) -> bool {
        let n = nums.len();
        let mut dp = vec![0; n+1];
        for k in 0..n {
            for i in 0..n-k {
                dp[i] = cmp::max(nums[i] - dp[i+1], nums[i+k] - dp[i])
            }
        }
        return dp[0] >= 0;
    }
}
